flink-runtime/src/test/java/org/apache/flink/runtime/rpc/heartbeat/HeartbeatSchedulerTest.java

```diff
@@ -0,0 +1,248 @@
+		);
+
+		TestingHeartbeatScheduler heartbeatScheduler = new TestingHeartbeatScheduler(sender, akkaRpcService, leaderSessionID,
+			targetGateway, "taskExecutor-test-address", "testTargetGateway", log, INITIAL_INTERVAL, INITIAL_TIMEOUT, MAX_TIMEOUT, DELAY_ON_ERROR, MAX_ATTEMPT_TIME);
+		heartbeatScheduler.start();
+
+		// verify heartbeat successful and syncHeartbeatResponse is invoked for sync heartbeat response
+		verify(sender, timeout(5000)).syncHeartbeatResponse(eq(response));
+		// verify heartbeat trigger is still on
+		Assert.assertFalse(heartbeatScheduler.isClosed());
+		heartbeatScheduler.close();
+	}
+
+	/**
+	 * test for fail upon first time, but successful after retry, and checking the heartbeatResponse is properly delivered
+	 * @throws Exception
+	 */
+	@Test
+	public void testHeartbeatRetrySuccessOnError() throws Exception {
+		HeartbeatSender sender = mock(HeartbeatSender.class);
+		UUID leaderSessionID = UUID.randomUUID();
+
+		HeartbeatReceiverGateway targetGateway = mock(HeartbeatReceiverGateway.class);
+
+		String response = "ok";
+		// triggerHeartbeat fails upon first time, but success on the second
+		when(targetGateway.triggerHeartbeat(any(UUID.class), any(FiniteDuration.class))).thenReturn(
+			Futures.<String> failed(new Exception("error happened")), // first attempt fail
+			Futures.successful(response) // second attempt success
+		);
+
+		TestingHeartbeatScheduler heartbeatScheduler = new TestingHeartbeatScheduler(sender, akkaRpcService, leaderSessionID,
+			targetGateway, "taskExecutor-test-address", "testTargetGateway", log, INITIAL_INTERVAL, INITIAL_TIMEOUT, MAX_TIMEOUT, DELAY_ON_ERROR, MAX_ATTEMPT_TIME);
+		heartbeatScheduler.start();
+
+		// verify heartbeat successful and syncHeartbeatResponse is invoked for sync heartbeat response
+		verify(sender, timeout(5000)).syncHeartbeatResponse(eq(response));
+		// verify heartbeat trigger is still on
+		Assert.assertFalse(heartbeatScheduler.isClosed());
+		heartbeatScheduler.close();
+	}
+
+	/**
+	 * test loss of heartbeat and check mark the resource failed after retry max attempt times,
+	 * @throws Exception
+	 */
+	@Test
+	public void testLostHeartbeatOnError() throws Exception {
+		HeartbeatSender sender = mock(HeartbeatSender.class);
+		UUID leaderSessionID = UUID.randomUUID();
+
+		HeartbeatReceiverGateway targetGateway = mock(HeartbeatReceiverGateway.class);
+
+		// triggerHeartbeat always fail
+		when(targetGateway.triggerHeartbeat(any(UUID.class), any(FiniteDuration.class))).thenReturn(
+			Futures.<String> failed(new Exception("error happened")));
+
+		TestingHeartbeatScheduler heartbeatScheduler = new TestingHeartbeatScheduler(sender, akkaRpcService, leaderSessionID,
+			targetGateway, "taskExecutor-test-address", "testTargetGateway", log, INITIAL_INTERVAL, INITIAL_TIMEOUT, MAX_TIMEOUT, DELAY_ON_ERROR, MAX_ATTEMPT_TIME);
+		heartbeatScheduler.start();
+		// verify lost heartbeat after max attempts and notifyLostHeartbeat is invoked
```

Comment:
yes, I define max timeout,  In this test, i also want to verify how many times sender will try to trigger heartbeat to receiver  in the max timeout period.  max attempt number = Math.ceil(MAX_ATTEMPT_TIMEOUT/DELAY_ON_ERROR) + 1))